common/lib/src/realm_types.dart

```diff
@@ -109,6 +97,37 @@ enum SyncErrorCategory {
   unknown,
 }
 
+/// Thrown or reporeted if an error occurs during synchronization
+/// {@category Sync}
+class SyncErrorCode extends RealmError {
+  /// The code of the error
+  final int code; // TODO: this should be an enum
+
+  /// The category of the error
+  final SyncErrorCategory category;
+
+  SyncErrorCode(String message, this.category, this.code) : super(message);
+}
+
+class SyncError extends RealmError {
+  final SyncErrorCode errorCode;
+  final bool isFatal;
+  final bool isUnrecognizedByClient;
+  final bool isClientResetRequested;
+  final Map<String, String> userInfoMap;
+  final int userInfoLength;
```

Comment:
We probably don't need those on the `SyncError` class.

* `isUnrecognizedByClient` does not bring much value to end users.
* If `isClientResetRequested` is `true`, we should throw a special error type that is different from the general-purpose `SyncError` and we should report it via the client reset handler, not the session error callback.
* `userInfoMap` and `userInfoLength` are internal and serve the purpose of propagating some client reset information - they should be translated into strongly typed properties.

Reply:
It is done as we decided - SessionError extends SyncError extends RealmError. Redundant properties are removed.